BarrysPeanutsTemporalJava/src/main/java/barryspeanuts/ShoppingCartWorkflowImpl.java

```diff
@@ -0,0 +1,98 @@
+package barryspeanuts;
+
+import barryspeanuts.model.Address;
+import barryspeanuts.model.CreditCard;
+import barryspeanuts.model.PurchaseItem;
+import io.temporal.workflow.Workflow;
+import io.temporal.workflow.WorkflowQueue;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+
+public class ShoppingCartWorkflowImpl implements ShoppingCartWorkflow {
+  private static final Logger logger = Workflow.getLogger(ShoppingCartWorkflowImpl.class);
+
+  boolean exit = false;
+  List<PurchaseItem> purchaseItems = new ArrayList<>();
+  private final WorkflowQueue<Runnable> queue = Workflow.newWorkflowQueue(1024);
+
+  @Override
+  public void startWorkflow() {
+    logger.info("Starting Workflow for Barry's Peanuts");
+    Workflow.await(() -> exit);
+  }
+
+  @Override
+  public List<PurchaseItem> queryPurchaseItems() {
+    return this.purchaseItems;
+  }
+
+  @Override
+  public void addItem(PurchaseItem purchaseItem) {
+    this.purchaseItems.add(purchaseItem);
+  }
+
+  @Override
+  public void addItems(List<PurchaseItem> purchaseItems) {
+    this.purchaseItems.addAll(purchaseItems);
+  }
+
+  @Override
+  public void removeItem(PurchaseItem purchaseItem) {
+    this.purchaseItems.remove(purchaseItem);
+  }
+
+  @Override
+  public void checkOut() {
+    logger.info("Checking out {} purchase items.", this.purchaseItems.toArray().length);
+  }
+
+  @Override
+  public void pay(Address billingAddress, CreditCard creditCard) {
+    for (PurchaseItem purchaseItem : this.purchaseItems) {
+      purchaseItem.setBillingAddress(Optional.ofNullable(billingAddress));
```

Comment:
Why mutate unaddressed items on pay? Just make the item immutable and if you need billing address somewhere or credit card somewhere, store.

Reply:
> Typically when I go to Amazon, I don't declare the billing address until time of payment. The same is try of declaring shipping address. I do this as the item is about to be shipped. Hence, the mutation.

Exactly, you don't declare billing address and shipping address on your items, you declare it on your payment. The Amazon items you selected in your cart are unchanged, as should these be. Payment/shipping info is unrelated to item.

Reply:
Addressed by refactoring